feat(vm-series): add new example with user data bootstrap method

damianfedeczko opened the pull request:
## Description

Example for vmseries with bootstrapping with user_data #111

## Motivation and Context

Add another example of the VM-Series module usage - this time using the User Data bootstrap method.

## How Has This Been Tested?

By running `terraform apply -var-file=example.tfvars`, like described in the "Usage" section of the README file.

## Screenshots (if appropriate)

Terraform deployment:
![image](https://user-images.githubusercontent.com/26055333/155088167-09f1cec6-860b-4775-b4b0-ffb13ec482d0.png)

SSH into the instance, check bootstrap status:
![image](https://user-images.githubusercontent.com/26055333/155104946-5aa4d4e9-b11c-46c8-b6af-388cd8f3ad28.png)

GUI access - disabled user/pass login:
![image](https://user-images.githubusercontent.com/26055333/155105461-15488574-5f12-493c-902b-0161ec74cb5a.png)

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes if appropriate.
- [ ] All new and existing tests passed.

migara reviewed (approved):
Great work @damianfedeczko 👍🏽 

Can you please change the example name to `standalone_vmseries_with_userdata_bootstrap` ?

damianfedeczko commented on examples/vmseries_userdata/example.tfvars:
```diff
@@ -0,0 +1,64 @@
+}
+
+# VM-Series
+ssh_key_name     = "dfedeczko-aws-lab"
```
> Great work @damianfedeczko 👍🏽
> 
> Can you please change the example name to `standalone_vmseries_with_userdata_bootstrap` ?

Sure thing 👍  Changed the ssh key name while you were reviewing to "CHANGE_ME", but will use the suggestion instead. Thanks!

michalbil commented on examples/standalone_vmseries_with_userdata_bootstrap/README.md:
```diff
@@ -0,0 +1,36 @@
+# Palo Alto Networks VM-Series NGFW Module Example for AWS Cloud
+
+A Terraform module example for deploying a VM-Series NGFW in AWS Cloud using the [User Data](https://docs.paloaltonetworks.com/vm-series/9-1/vm-series-deployment/bootstrap-the-vm-series-firewall/choose-a-bootstrap-method.html#idf6412176-e973-488e-9d7a-c568fe1e33a9_id3433e9c0-a589-40d5-b0bd-4bc42234aa0f) bootstrap method.
```
```suggestion
A Terraform module example for deploying a VM-Series NGFW in AWS Cloud using the [User Data](https://docs.paloaltonetworks.com/vm-series/10-0/vm-series-deployment/bootstrap-the-vm-series-firewall/choose-a-bootstrap-method.html#idf6412176-e973-488e-9d7a-c568fe1e33a9_id3433e9c0-a589-40d5-b0bd-4bc42234aa0f) bootstrap method.
```
10.0 is a default version for vmseries module, so let's refer to 10.0 docs.

damianfedeczko commented:
@migara @michalbil - I've updated the module `variables` and README.md, and added the pre-commit webhook to the example README. Once Michal merges the `outputs` change into the VM-Series module, I'll update the one in the example accordingly. Michal - please do not run the pre-commit webhook for your change - I'll do that for both of us. Thanks!

damianfedeczko commented:
@migara @michalbil - I've merged Michal's changes into `develop` and adjusted this branch accordingly - please review - and let's merge if thats fine with you.

michalbil commented on modules/vmseries/variables.tf:
```diff
@@ -118,6 +118,10 @@ variable "bootstrap_options" {
   For more information on available methods, please refer to VM-Series documentation for specific version.
   For 10.0 docs are available [here](https://docs.paloaltonetworks.com/vm-series/10-0/vm-series-deployment/bootstrap-the-vm-series-firewall.html).
+
+  If you need to supply more than one bootstrap option, you will need to use a map instead of a string in your `.tvfars` file.
```
@damianfedeczko this is not accurate, this input will only accept strings. Any info on how this can be handled on calling module level should not be included in this var's description.

damianfedeczko replied:
Ok, so please update the variable description accordingly - I don't have a better idea how to handle that. Using a string isn't a real-world situation, for most of the cases, we use a map with multiple options.

michalbil replied:
Suggestion added.

As for real world example of string-only `bootstrap_options` - please consider changes as below for the existing tgw example:
- tfvars file (bootstrap options as string, not map)
```hcl
[...]
vmseries_common = {
  vmseries_version = "10.1.3"
  bootstrap_options = <<-EOT
  mgmt-interface-swap=enable
  plugin-op-commands=aws-gwlb-inspect:enable,numa-perf-optimize:enable
  type=dhcp-client
  EOT
}
[...]
```
- `vmseries.tf` (pass that string directly to module input)
```hcl
[...]
bootstrap_options = var.vmseries_common.bootstrap_options
[...]
```
This results in user_data being populated properly.

Accepting data that can be directly passed into `aws_instance.user_data` at vmseries module level (without any transformations inside) was one of requirements raised under both PRs with vmseries refactor.

michalbil replied:
@damianfedeczko I've updated the PR as discussed, checks on the firewall:
```
admin@vms01> show system bootstrap status 

Bootstrap Phase               Status         Details
===============               ======         =======
Media Detection               Success        Media detected successfully
Media Sanity Check            Success        Media sanity check successful
VM-Series PlG OP CMDS         Success        Operational command enable aws-gwlb-inspect completed successfully; Operational command enable aws-gwlb-overlay-routing completed successfully
Parsing of Initial Config     Successful     
Auto-commit                   Successful     

admin@vms01> show plugins vm_series aws gwlb 

    GWLB enabled    :    True
    Overlay Routing :    True
    ================================================
    VPC endpoint              Interface      
    ================================================
```

damianfedeczko replied:
Great 💪

damianfedeczko commented:
@migara can we merge this PR now?
